L1Trigger/L1TGlobal/src/GlobalBoard.cc

```diff
@@ -388,35 +378,36 @@ void l1t::GlobalBoard::receiveMuonShowerObjectData(const edm::Event& iEvent,
                                      << "\nrequested in configuration, but not found in the event.\n";
       }
     } else {
-      //Loop over Muon Showers in this bx
+      // Loop over Muon Showers in this bx
       int nObj = 0;
       for (auto mu = muonData->begin(0); mu != muonData->end(0); ++mu) {
         if (nObj < nrL1MuShower) {
-          /* Important here to split up the single object into 4 separate MuonShower
-             bits for the global board. This is because the UTM library considers those bits separate as well
+          /* NOTE: here the single object is split up into 4 separate MuonShower objects 
+	     similarly to the description in the UTM library, where the conditions are four different objects.
            */
-          l1t::MuonShower mus0;
-          l1t::MuonShower mus1;
-          l1t::MuonShower musOutOfTime0;
-          l1t::MuonShower musOutOfTime1;
-
-          mus0.setMus0(mu->mus0());
-          mus1.setMus1(mu->mus1());
-          musOutOfTime0.setMusOutOfTime0(mu->musOutOfTime0());
-          musOutOfTime1.setMusOutOfTime1(mu->musOutOfTime1());
-
-          (*m_candL1MuShower).push_back(0, &mus0);
-          (*m_candL1MuShower).push_back(0, &mus1);
-          (*m_candL1MuShower).push_back(0, &musOutOfTime0);
-          (*m_candL1MuShower).push_back(0, &musOutOfTime1);
+          l1t::MuonShower* musOneNominalInTime = new l1t::MuonShower(false, false, false, false, false, false);
+          l1t::MuonShower* musOneTightInTime = new l1t::MuonShower(false, false, false, false, false, false);
+          l1t::MuonShower* musOutOfTime0 = new l1t::MuonShower(false, false, false, false, false, false);
+          l1t::MuonShower* musOutOfTime1 = new l1t::MuonShower(false, false, false, false, false, false);
+
+          musOneNominalInTime->setOneNominalInTime(mu->isOneNominalInTime());
+          musOneTightInTime->setOneTightInTime(mu->isOneTightInTime());
+          musOutOfTime0->setMusOutOfTime0(mu->musOutOfTime0());
+          musOutOfTime1->setMusOutOfTime1(mu->musOutOfTime1());
+
+          (*m_candL1MuShower).push_back(0, &(*musOneNominalInTime));
+          (*m_candL1MuShower).push_back(0, &(*musOneTightInTime));
+          (*m_candL1MuShower).push_back(0, &(*musOutOfTime0));
+          (*m_candL1MuShower).push_back(0, &(*musOutOfTime1));
+
```

Comment:
I assume that using a `std::unique_ptr` is not possible here because of the .h function interface here?
https://github.com/cms-sw/cmssw/blob/29ecd54f381cef5cd361d882d8bfc9267cdf56ee/L1Trigger/L1TGlobal/interface/GlobalBoard.h#L153

If it is, I think that might be better, because I am concerned that the allocated `musOneNominalInTime`, `musOneTightInTime`,`musOutOfTime0`, & `musOutOfTime1` may be leaking memory. They are put into the BXVector `m_candL1MuShower` but I don't see anything that explicitly de-allocates this allocated memory, and I don't see anything inside the BXVector code that would do it either. 

Later `clear` is called here:

https://github.com/cms-sw/cmssw/blob/29ecd54f381cef5cd361d882d8bfc9267cdf56ee/L1Trigger/L1TGlobal/src/GlobalBoard.cc#L1094

but this more or less boils down to  calling a `std::vector.clear()` I think, which may call destructors, but is not equivalent to `delete` on the objects to the best of my knowledge. There is also BX range setting here: 

https://github.com/cms-sw/cmssw/blob/29ecd54f381cef5cd361d882d8bfc9267cdf56ee/L1Trigger/L1TGlobal/src/GlobalBoard.cc#L1095

But I don't think that has an effect if the BX numbers are the same as before, (isn't -2,-1,0,1,2 pretty typical for a BX vector?)

https://github.com/cms-sw/cmssw/blob/08111d0a8317a22ad54d49767e6a999b5ae19149/DataFormats/L1Trigger/interface/BXVector.icc#L20-L45

Perhaps I am missing something though. Is there something I am not seeing that makes sure that the memory allocated for `musOneNominalInTime`, `musOneTightInTime`,`musOutOfTime0`, and `musOutOfTime1` is later de-allocated?

Reply:
Hi @aloeliger , these changes were mostly done by me rather quickly. I didn't try using `std::unique_ptr`. This is only one possible fix to this issue. Anybody who is more of an expert on uGT emulator should look into this to see if there are more efficient solutions. I'm also not a huge fan of splitting one object to four objects which creates this problem to begin with. Other uGT objects are just pushed back to the vector just by their iterator.